### PR TITLE
zstd image transport plugin

### PR DESCRIPTION
**Pull request by ahcorde**
zstd image transport plugin

### CODE REVIEW

File: zstd_image_transport/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+find_package(ament_cmake REQUIRED)
+find_package(image_transport REQUIRED)
+find_package(pluginlib REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(sensor_msgs REQUIRED)
```
Comment (ijnek) [inline zstd_image_transport/CMakeLists.txt:18]:
```suggestion
find_package(sensor_msgs REQUIRED)
find_package(ZLIB REQUIRED)
```

File: zstd_image_transport/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+add_library(
+  ${PROJECT_NAME} SHARED
+  src/zlib_cpp.cpp
+  src/zstd_publisher.cpp
+  src/zstd_subscriber.cpp
+  src/manifest.cpp
+)
+
+target_link_libraries(${PROJECT_NAME})
```
Comment (ijnek) [inline zstd_image_transport/CMakeLists.txt]:
```suggestion
target_link_libraries(${PROJECT_NAME} ZLIB::ZLIB)
```

File: zstd_image_transport/package.xml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  </description>
+  <maintainer email="ahcorde@gmail.com">Alejandro Hernandez Cordero</maintainer>
+  <license>BSD</license>
+
+  <url type="website">http://www.ros.org/wiki/image_transport_plugins</url>
+  <author>Alejandro Hernandez Cordero</author>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+
+  <depend>image_transport</depend>
```
Comment (ijnek) [inline zstd_image_transport/package.xml:16]:
```suggestion
  <depend>image_transport</depend>
  <depend>zlib</depend>
```

File: zstd_image_transport/src/zlib_cpp.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  do {
+    // Prepare outcoming buffer and size.
+    zs_.avail_out = MAX_CHUNK_SIZE;
+    zs_.next_out = out_buffer;
+    // Decompress data.
+    ret = inflate(&zs_, Z_NO_FLUSH);
+    switch (ret) {
+      case Z_NEED_DICT:
+        // Incoming data is invalid.
+        return std::move(out_data_list);
```
Comment (ijnek) [inline zstd_image_transport/src/zlib_cpp.cpp]: ``warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]``
Let the compiler do return value optimization instead by not using move semantics.

```suggestion
        return out_data_list;
```

File: zstd_image_transport/src/zlib_cpp.cpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      case Z_DATA_ERROR:
+      case Z_MEM_ERROR:
+        // Critical error.
+        return std::move(out_data_list);
```
Comment (ijnek) [inline zstd_image_transport/src/zlib_cpp.cpp]:
```suggestion
        return out_data_list;
```

File: zstd_image_transport/src/zlib_cpp.cpp [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    // Outcome size.
+    std::size_t out_size = MAX_CHUNK_SIZE - zs_.avail_out;
+    // Allocate outcome buffer.
+    std::shared_ptr<DataBlock> out_data = AllocateData(out_size);
+    memcpy(out_data->ptr, out_buffer, out_size);
+    out_data_list.push_back(std::move(out_data));
+  } while (zs_.avail_out == 0);
+  return std::move(out_data_list);
```
Comment (ijnek) [inline zstd_image_transport/src/zlib_cpp.cpp]:
```suggestion
  return out_data_list;
```

File: zstd_image_transport/src/zstd_publisher.cpp [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+  compressed.header = message.header;
+  compressed.format = "zstd";
+  publish_fn(compressed);
+}
+
+void ZstdPublisher::declareParameter(
+  const std::string & base_name,
+  const ParameterDefinition & definition)
+{
+  // transport scoped parameter (e.g. image_raw.compressed.format)
```
Comment (ijnek) [inline zstd_image_transport/src/zstd_publisher.cpp]:
```suggestion
  // transport scoped parameter (e.g. image_raw.zstd.zstd_level)
```

**Review by ijnek — CHANGES_REQUESTED** [review]
@ahcorde First of all, I apologise for the super long delay in getting this reviewed. Thank you for this addition. The changes look solid, and I have tested it and they work great (and it's fast!).

I have two minor request for changes, and I think otherwise everything looks good.
I had a look at the CI, and I don't see any tests failing (but rather some errors like):

```
03:10:45 [GNU C Compiler (gcc)] [-ERROR-] - '/tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: /tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp
03:10:45 [GNU C Compiler (gcc)] [-ERROR-] - '/tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: /tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp
03:10:45 [GNU C Compiler (gcc)] [-ERROR-] - '/tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp', IO exception has been thrown: java.nio.file.NoSuchFileException: /tmp/ws/src/image_transport_plugins/zstd_image_transport/src/zlib_cpp.cpp
```
I've seen them infrequently show up on the buildfarm, but not quite sure what the problem is.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20094/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14541/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20814/)

**Review by ijnek — APPROVED** [review]

**Comment by authaldo** [comment]
Small question regarding the new plugin: 
Has the `<exec_depend>zstd_image_transport</exec_depend>` within the package.xml of the metapackage been omitted on purpose?

**Comment by ahcorde** [comment]
@authaldo I forgot to include it. PR https://github.com/ros-perception/image_transport_plugins/pull/162

**Pull request merged**